src/app/qgsversionmigration.cpp

```diff
@@ -0,0 +1,222 @@
+/***************************************************************************
+  qgsversionmigration.cpp - QgsVersionMigration
+
+ ---------------------
+ begin                : 30.7.2017
+ copyright            : (C) 2017 by nathan
+ email                : woodrow.nathan at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#include "qgsversionmigration.h"
+#include "qgssettings.h"
+#include "qgslogger.h"
+#include "qsettings.h"
+#include "qgsmessagelog.h"
+#include "qgsapplication.h"
+#include "qgssymbol.h"
+#include "qgsstyle.h"
+#include "qgssymbollayerutils.h"
+#include "qgsreadwritecontext.h"
+
+#include <QFile>
+#include <QTextStream>
+#include <QDir>
+#include <QSqlDatabase>
+#include <QSqlQuery>
+#include <QSqlError>
+#include <QDomDocument>
+
+QgsVersionMigration::QgsVersionMigration()
+{
+
+}
+
+QgsVersionMigration::~QgsVersionMigration()
+{
+
+}
+
+QgsVersionMigration *QgsVersionMigration::canMigrate( int fromVersion, int toVersion )
+{
+  if ( fromVersion == 20000 && toVersion >= 29900 )
+  {
+    return new Qgs2To3Migration();
+  }
+  return nullptr;
+}
+
+QgsError Qgs2To3Migration::runMigration()
+{
+  QgsError error;
+  QgsError settingsErrors = migrateSettings();
+  if ( !settingsErrors.isEmpty() )
+  {
+    // TODO Merge error messages
+  }
+  QgsError stylesError = migrateStyles();
+  return error;
+}
+
+bool Qgs2To3Migration::requiresMigration()
+{
+  QgsSettings settings;
+  bool alreadyMigrated = settings.value( QStringLiteral( "migration/settings" ), false ).toBool();
+  int  settingsMigrationVersion = settings.value( QStringLiteral( "migration/fileVersion" ), 0 ).toInt();
+  QFile migrationFile( migrationFilePath() );
+  if ( migrationFile.open( QIODevice::ReadOnly | QIODevice::Text ) )
+  {
+    QTextStream in( &migrationFile );
+    QString line = in.readLine();
+    if ( line.startsWith( "#" ) && line.contains( QStringLiteral( "version=" ) ) )
+    {
+      QStringList parts = line.split( '=' );
+      mMigrationFileVersion = parts.at( 1 ).toInt();
```

Comment:
better be safe here and check the length of parts first - otherwise we potentially could hit a situation where qgis crashes on launch with a malformed migration file